<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
<title>Exception: <code>PyException</code></title>
<!-- 2016-03-09 周三 19:48 -->
<meta  http-equiv="Content-Type" content="text/html;charset=utf-8" />
<meta  name="generator" content="Org-mode" />
<meta  name="author" content="Chen Zhou" />
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center; }
  .todo   { font-family: monospace; color: red; }
  .done   { color: green; }
  .tag    { background-color: #eee; font-family: monospace;
	    padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  pre.src-sh:before    { content: 'sh'; }
  pre.src-bash:before  { content: 'sh'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-R:before     { content: 'R'; }
  pre.src-perl:before  { content: 'Perl'; }
  pre.src-java:before  { content: 'Java'; }
  pre.src-sql:before   { content: 'SQL'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.right  { text-align: center;  }
  th.left   { text-align: center;   }
  th.center { text-align: center; }
  td.right  { text-align: right;  }
  td.left   { text-align: left;   }
  td.center { text-align: center; }
  dt { font-weight: bold; }
  .footpara:nth-child(2) { display: inline; }
  .footpara { display: block; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  /*]]>*/-->
</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2013 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Exception: <code>PyException</code></h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#sec-1">1. The header file</a></li>
<li><a href="#sec-2">2. Details</a></li>
</ul>
</div>
</div>

<div id="outline-container-sec-1" class="outline-2">
<h2 id="sec-1"><span class="section-number-2">1</span> The header file</h2>
<div class="outline-text-2" id="text-1">
<p>
The <code>PyException</code> object is inherited from <code>PyObject</code>.
</p>

<div class="org-src-container">

<pre class="src src-c++">#ifndef PYEXCEPTION_H_
#define PYEXCEPTION_H_

#include "PyObject.h"
class PyFrame;
#include &lt;string&gt;
using namespace std;

class PyException : public PyObject {
public:
    PyException(int exceptType, PyObject* val);
    PyException(int exceptType, string message);
    virtual ~PyException();
    int getExceptionType();
    void tracebackAppend(PyFrame* frame);
    string toString();
    PyType* getType();
    PyObject* getTraceBack();
    void printTraceBack();

protected:
    int exceptionType;
    PyObject* val;
    vector&lt;PyFrame*&gt; traceback;
    virtual PyObject* __excmatch__(vector&lt;PyObject*&gt;* args);
};
</pre>
</div>

<p>
The constructor of class <code>PyException</code> takes an int which stands
for a exception type and either a pointer to <code>PyObject</code> or a string
message.
</p>

<p>
To provide useful error massages, the exception class has to keep
records of the frames from which an exception may be invoked. This
function is provided mainly by three functions: <code>tracebackAppend</code>
which takes a pointer to a frame, <code>getTraceBack</code>, and
<code>printTraceBack</code>. They in turn work with the protected field members:
<code>exceptionType</code>, <code>val</code> (a pointer to <code>PyObject</code>), <code>traceback</code> (a
vector of pointers to <code>PyFrame</code>), and a virtual function
<code>__excmatch__</code> which checks the argument passed to it against the type
of the current object's exception. However, their definitions remain to
be revealed in later sections about the <code>PyException.cpp</code> file.
</p>

<div class="org-src-container">

<pre class="src src-c++">const int PYEXCEPTION = 1;
const int PYEMPTYSTACKEXCEPTION = 2;
const int PYPARSEEXCEPTION = 3;
const int PYILLEGALOPERATIONEXCEPTION = 4;
const int PYWRONGARGCOUNTEXCEPTION = 5;
const int PYSTOPITERATIONEXCEPTION = 6;
const int PYMATCHEXCEPTION = 7;

#endif
</pre>
</div>

<p>
Finally, the header file defines seven exception types with integers
one to seven.
</p>

<p>
Using <code>int</code> instead of <code>enum</code> to tag seven exception types is for the
reason that later these integers are used as keys to make a map
representing exceptions.
</p>
</div>
</div>

<div id="outline-container-sec-2" class="outline-2">
<h2 id="sec-2"><span class="section-number-2">2</span> Details</h2>
<div class="outline-text-2" id="text-2">
<p>
At first, in <code>PyException.cpp</code> a <code>struct</code> is defined to store names of
exception. This <code>struct</code> also contains a method which create an
<code>unordered_map</code> whose keys are defined in the header file, the values
are string representations.
</p>

<div class="org-src-container">

<pre class="src src-c++">#include "PyList.h"
#include "PyBool.h"
#include &lt;unordered_map&gt;
#include &lt;sstream&gt;

using namespace std;

struct names {

    static unordered_map&lt;int, string&gt; create_map() {
	unordered_map&lt;int, string&gt; m;
	m[PYEXCEPTION] = "Exception";
	m[PYEMPTYSTACKEXCEPTION] = "EmptyStackException";
	m[PYPARSEEXCEPTION] = "ParseException";
	m[PYILLEGALOPERATIONEXCEPTION] = "IllegalOperationException";
	m[PYWRONGARGCOUNTEXCEPTION] = "WrongArgCountExeption";
	m[PYSTOPITERATIONEXCEPTION] = "StopIterationException";
	m[PYMATCHEXCEPTION] = "MatchException";

	return m;
    }
};

static unordered_map&lt;int, string&gt; excnames = names::create_map();
</pre>
</div>

<p>
Since there are two versions of constructors, one needs a pointer to
<code>PyObject</code>, another takes a message string, the implementation of
constructor also has two parts. If a <code>PyException</code> object is
constructed by a pointer to <code>PyObject</code>, the private field value <code>val</code>
would be a pointer. If it is a string, this string would be stored into
the field <code>val</code>.
</p>

<div class="org-src-container">

<pre class="src src-c++">PyException::PyException(int exception, PyObject* v):
    PyObject(), exceptionType(exception), val(v)
{
    dict["__excmatch__"] =
	(PyObject* (PyObject::*)(vector&lt;PyObject*&gt;*))
	(&amp;PyException::__excmatch__);
}

PyException::PyException(int exception, string msg):
    PyObject(), exceptionType(exception), val(new PyStr(msg))
{
    dict["__excmatch__"] =
	(PyObject* (PyObject::*)(vector&lt;PyObject*&gt;*))
	(&amp;PyException::__excmatch__);
}
</pre>
</div>

<p>
The two constructors respectively add a magic method <code>__excmatch__</code> to
their private member <code>dict</code>. The value of <code>__excmatch__</code> in this map
is the private function <code>__excmatch__</code> declared in the header file.
</p>

<p>
The descructor would delete its <code>val</code> value.
</p>

<div class="org-src-container">

<pre class="src src-c++">PyException::~PyException() {
    try {
	delete val;
    } catch (...) {}
}
</pre>
</div>

<p>
The function <code>getExceptionType</code> is just a help function to get value
from the private field <code>exceptionType</code> which is not accessible
directly. Setting <code>exceptionType</code> as a private member can prevent
other elements of this code changing this property accidentally, which
could be a disaster for debugging.
</p>

<div class="org-src-container">

<pre class="src src-c++">int PyException::getExceptionType() {
    return exceptionType;
}

string PyException::toString() {
    return val-&gt;toString();
}

PyType* PyException::getType() {
    return PyTypes[PyExceptionTypeId];
}
</pre>
</div>

<p>
The class <code>PyException</code> has its own version of member function
<code>toString</code> (this function is defined as virtual in the parent class
<code>PyObject</code>). It do not provide all information about this object, just
invoke the <code>val</code>'s <code>toString</code> method instead. At this stage, the <code>val</code>
in the <code>PyException</code> is a pointer to either a <code>PyStr</code> or a
<code>PyObject</code>. Both of them have their own <code>toString</code> method.
</p>

<p>
The member function <code>getType</code> of <code>PyException</code> is also virtual, the
return value for this specific object is a pointer to <code>PyType</code> object
which is found by searching a <code>PyExceptionTypeId</code> (defined in <code>PyType</code>
library) in an <code>unordered_map</code> defined in <code>main.cpp</code>.
</p>

<p>
The <code>trackback</code> field is managed by three member functions <code>printTraceBack</code>,
<code>tracebackAppend</code> and <code>getTraceback</code>.
</p>

<div class="org-src-container">

<pre class="src src-c++">void PyException::printTraceBack() {
    for (int k=0; k&lt;traceback.size(); k++) {
	cerr &lt;&lt; "=========&gt; At PC="
	     &lt;&lt; (traceback[k]-&gt;getPC()-1)
	     &lt;&lt; " in this function."
	     &lt;&lt; endl;
	cerr &lt;&lt; traceback[k]-&gt;getCode().prettyString("", true);
    }
}

void PyException::tracebackAppend(PyFrame* frame) {
    traceback.push_back(frame);
}

PyObject* PyException::getTraceBack() {
    return new PyList((vector&lt;PyObject*&gt;*)&amp;traceback);
}
</pre>
</div>

<p>
The function <code>printTraceBack</code> would loop through all elements of the
<code>tracebak</code> vector and print each element's <code>PC</code> and code
representation. The vector <code>traceback</code> consists of pointers to the
object <code>PyFrame</code>. Appending a new frame to the <code>traceback</code> naturally
becomes pushing a new element to that vector. The member function
<code>getTraceBack</code> cast the pointer of the <code>traceback</code> field to a pointer
to a vector of <code>PyObject</code> pointers, then return them in a <code>PyList</code>
object.
</p>

<p>
The magic function the constructor added to every <code>PyException</code> object
compares <code>this</code> exception object against its argument.
</p>

<div class="org-src-container">

<pre class="src src-c++">PyObject* PyException::__excmatch__(vector&lt;PyObject*&gt;* args) {
    ostringstream msg;

    if (args-&gt;size() != 1) {
	msg &lt;&lt; "TypeError: excepted 1 arguments, got "
	    &lt;&lt; args-&gt;size();
	throw new PyException(PYWRONGARGCOUNTEXCEPTION, msg.str());
    }

    PyObject* arg = (*args)[0];

    if (this-&gt;getType() == arg)
	return new PyBool(true);

    if (this-&gt;getType() != arg-&gt;getType()) {
	msg &lt;&lt; "TypeError: Exception match type mismatch. Excepted Exception Object got "
	    &lt;&lt; arg-&gt;toString();
	throw new PyException(PYILLEGALOPERATIONEXCEPTION, msg.str());
    } // The official version of this code has not this pair of parenthesis.

    PyException* other = (PyException*) arg;

    return new PyBool(this-&gt;getExceptionType() == other-&gt;getExceptionType());
}
</pre>
</div>

<p>
The operation this function taken can be split to three steps. First,
check the number of argument(s), if it does not equal to one, an error
would be thrown. Second, if the types of this object and the argument
do not agree, an error thrown too. The last step returns the outcome of
comparing values of <code>getExceptionType</code>.
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Chen Zhou</p>
<p class="date">Created: 2016-03-09 周三 19:48</p>
<p class="creator"><a href="http://www.gnu.org/software/emacs/">Emacs</a> 25.1.50.2 (<a href="http://orgmode.org">Org</a> mode 8.2.10)</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
